reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,486 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.vaticle.typedb.core.reasoner.planner;
+
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+import com.vaticle.typedb.core.common.iterator.FunctionalIterator;
+import com.vaticle.typedb.core.common.parameters.Label;
+import com.vaticle.typedb.core.logic.LogicManager;
+import com.vaticle.typedb.core.logic.Rule;
+import com.vaticle.typedb.core.logic.resolvable.Concludable;
+import com.vaticle.typedb.core.logic.resolvable.Resolvable;
+import com.vaticle.typedb.core.logic.resolvable.ResolvableConjunction;
+import com.vaticle.typedb.core.logic.resolvable.Unifier;
+import com.vaticle.typedb.core.pattern.constraint.thing.RelationConstraint;
+import com.vaticle.typedb.core.pattern.variable.ThingVariable;
+import com.vaticle.typedb.core.pattern.variable.TypeVariable;
+import com.vaticle.typedb.core.pattern.variable.Variable;
+import com.vaticle.typedb.core.traversal.common.Identifier;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.vaticle.typedb.common.collection.Collections.list;
+import static com.vaticle.typedb.common.collection.Collections.set;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class AnswerCountEstimator {
+    private final LogicManager logicMgr;
+    private final Map<ResolvableConjunction, ConjunctionAnswerCountEstimator> estimators;
+
+    // Cycle handling
+    private final ArrayList<ResolvableConjunction> initializationStack;
+    private final Set<ResolvableConjunction> cycleHeads;
+    private final Set<ResolvableConjunction> toReset;
+    private final AnswerCountModel answerCountModel;
+
+    public AnswerCountEstimator(LogicManager logicMgr) {
+        this.logicMgr = logicMgr;
+        this.answerCountModel = new AnswerCountModel(this);
+        this.estimators = new HashMap<>();
+        this.initializationStack = new ArrayList<>();
+        this.cycleHeads = new HashSet<>();
+        this.toReset = new HashSet<>();
+    }
+
+    public long estimateAllAnswers(ResolvableConjunction conjunction) {
+        registerConjunctionAndInitialize(conjunction);
+        return estimators.get(conjunction).estimateAllAnswers();
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter) {
+        return estimateAnswers(conjunction, variableFilter, logicMgr.compile(conjunction));
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter, Set<Resolvable<?>> includedResolvables) {
```

Comment:
finer detail:
the order of the arguments would be better done as `Conjunction, Resolvables, Filter`.
Because we tend to read left to right, it makes more sense to think of the arguments as "estimate answers from this conjunction within these resolvables then narrowed to this filter". The other way around, "estimate answers from this conjunction, narrows to this filter, then within these resolvables" is conceptually reversing the dependencies

Reply:
Does this still hold when you have two variants of the same method, where one has an extra argument?
This is the equivalent of having a default argument value in languages which support it. Such languages often require arguments with default values to be after all arguments without default values.

Reply:
it can get tricky with overloading to be honest. It's a good principle but sometimes the "extra argument" must come at the end to maintain symmetry with the other versions of the method:
`run(A, B)`
`run(A,B,C)` -- this version C would actually read better before `B` but we prioritise the symmetry with the other version of this method